### PR TITLE
Mark messages as seen when parent thread is clicked on

### PR DESCRIPTION
**Pull request by KevinBoulongne**

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/data/api/ApiRepository.kt [hunk 1]
```diff
@@ -79,11 +79,11 @@ object ApiRepository : ApiRepositoryCore() {
...
         callApi(ApiRoutes.quotas(mailboxMailbox, mailboxHostingId), GET)
 
-    // fun markAsSeen(mailboxUuid: String, messages: ArrayList<Message>): ApiResponse<ArrayList<Seen>> =
-    //     callApi(ApiRoutes.messageSeen(mailboxUuid), POST, mapOf("uids" to messages.map { it.uid }))
+     fun markMessagesAsSeen(mailboxUuid: String, messages: ArrayList<Message>): ApiResponse<ArrayList<Seen>> =
+         callApi(ApiRoutes.messageSeen(mailboxUuid), POST, mapOf("uids" to messages.map { it.uid }))
 
-    // fun markAsUnseen(mailboxUuid: String, messages: ArrayList<Message>): ApiResponse<ArrayList<Seen>> =
-    //     callApi(ApiRoutes.messageUnseen(mailboxUuid), POST, mapOf("uids" to messages.map { it.uid }))
+     fun markMessagesAsUnseen(mailboxUuid: String, messages: ArrayList<Message>): ApiResponse<ArrayList<Seen>> =
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/api/ApiRepository.kt:85]: Can you update `ArrayList` please ?

File: app/src/main/java/com/infomaniak/mail/data/api/ApiRepository.kt [hunk 2]
```diff
@@ -79,11 +79,11 @@ object ApiRepository : ApiRepositoryCore() {
     fun getQuotas(mailboxHostingId: Int, mailboxMailbox: String): ApiResponse<Quotas> =
         callApi(ApiRoutes.quotas(mailboxMailbox, mailboxHostingId), GET)
 
-    // fun markAsSeen(mailboxUuid: String, messages: ArrayList<Message>): ApiResponse<ArrayList<Seen>> =
-    //     callApi(ApiRoutes.messageSeen(mailboxUuid), POST, mapOf("uids" to messages.map { it.uid }))
+     fun markMessagesAsSeen(mailboxUuid: String, messages: ArrayList<Message>): ApiResponse<ArrayList<Seen>> =
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/api/ApiRepository.kt:82]: Can you update `ArrayList` please ?

File: app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt [hunk 3]
```diff
@@ -95,6 +95,9 @@ object MailboxContentController {
         MailRealm.mailboxContent.writeBlocking { getLatestMessage(uid)?.let(::delete) }
     }
 
+    fun getLatestMessage(uid: String): Message? =
+        MailRealm.mailboxContent.writeBlocking { getLatestMessage(uid) }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt:99]: Why `getLatestMessage` is in `writeBlocking ` ?

File: app/src/main/java/com/infomaniak/mail/data/models/thread/Thread.kt [hunk 4]
```diff
@@ -85,6 +85,18 @@ class Thread : RealmObject {
...
+        MailRealm.mailboxContent.writeBlocking {
+            MailboxContentController.getThread(uid)?.let { coldThread ->
+                findLatest(coldThread)?.let { hotThread ->
+                    hotThread.messages.forEach { it.seen = true }
+                    hotThread.unseenMessagesCount = 0
+                    ApiRepository.markMessagesAsSeen(mailboxUuid, ArrayList(messages))
+                }
+            }
+        }
+    }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/models/thread/Thread.kt:98]: Please move in a `ViewModel`, and update `ArrayList`

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt [hunk 5]
```diff
@@ -35,14 +37,17 @@ class ThreadViewModel : ViewModel() {
 
     fun getMessagesFromRealmThenFetchFromApi(threadUid: String): List<Message> {
         val thread = readThreadFromRealm(threadUid)
+        thread?.select()
+        MailRealm.currentMailboxObjectIdFlow.value?.let {
+            thread?.markAsSeen(MailboxInfoController.getMailbox(it)?.uuid ?: "")
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt:42]: Useless to call this if `getMailbox` return null

**Review by sirambd — CHANGES_REQUESTED** [review]
Check Request changes, and (update + refactor), if necessary

File: app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt (hunk 3 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt:99]: There is 2 `getLatestMessage`:
- the public one
- the private one that is an extension function of `MutableRealm`, so it needs to be called in a `write` block.

Comment (sirambd) [reply app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt:99]: okay 👍🏾

**Review by sirambd — APPROVED** [review]

**Review by JorisBodin — APPROVED** [review]

**Pull request merged**